src/Compilers/VisualBasic/Portable/CodeGen/EmitExpression.vb

```diff
@@ -1034,7 +1034,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGen
                     ' if the receiver is actually a value type it would need to be boxed.
                     ' let .constrained sort this out. 
 
-                    callKind = If(receiverType.IsReferenceType AndAlso Not AllowedToTakeRef(receiver, AddressKind.ReadOnly),
+                    Debug.Assert(Not receiverType.IsReferenceType OrElse receiver.Kind <> BoundKind.ComplexConditionalAccessReceiver)
+                    callKind = If(receiverType.IsReferenceType AndAlso
+                                   (receiver.Kind = BoundKind.ConditionalAccessReceiverPlaceholder OrElse Not AllowedToTakeRef(receiver, AddressKind.ReadOnly)),
```

Comment:
>AllowedToTakeRef [](start = 110, length = 16)

I am a bit surprised you didn't fix this by making `AllowedToTakeRef` return false for a `ConditionalAccessReceiverPlaceholder`

Reply:
Unless one is really allowed to take a ref of the placeholder, I would think that is a more correct fix.

---
In reply to: [152643217](https://github.com/dotnet/roslyn/pull/23359#discussion_r152643217) [](ancestors = 152643217)

Reply:
> I am a bit surprised you didn't fix this by making AllowedToTakeRef return false for a ConditionalAccessReceiverPlaceholder

> Unless one is really allowed to take a ref of the placeholder, I would think that is a more correct fix.

The placeholder node simply indicates that whatever we need is already on the stack in a suitable form. The logic I added ensures that we respect the form of the receiver that EmitConditionalAccess puts on the stack. Changing behavior of AllowedToTakeRef is going to break other scenarios, when a type parameter is not known to be a reference, or a value type. For example, when there are no constraints at all, or all constraints are interfaces, etc.

Reply:
When receiver is known to be a reference type CallVirt is better than ConstrainedCallVirt because the result would be the same but the second requires a reference. I.E. if you have a value on the left, you would need to store it into a temp and push a ref to that temp.

C# is even more aggressive here with use of CallVirt - if receiver is not already a ref (only few cases are possible - like a result of a ref method invocation), then it forces CallVirt. 
In particular the receiver of a conditional call is never a ref when it has a reference type - since we need to null-check the value.

I think the fix is good enough. 
Porting `private bool IsRef(BoundExpression receiver)` from C# could have the same effect and improve code in related cases.

---
In reply to: [152643493](https://github.com/dotnet/roslyn/pull/23359#discussion_r152643493) [](ancestors = 152643493,152643217)

Reply:
> Porting private bool IsRef(BoundExpression receiver) from C# could have the same effect and improve code in related cases.

Opened https://github.com/dotnet/roslyn/issues/23362